lfsapi/endpoint_finder.go

```diff
@@ -117,14 +119,53 @@ func (e *endpointGitFinder) RemoteEndpoint(operation, remote string) lfshttp.End
 		return e.NewEndpoint(operation, url)
 	}
 
-	// finally fall back on git remote url (also supports pushurl)
+	// fall back on git remote url (also supports pushurl)
 	if url := e.GitRemoteURL(remote, operation == "upload"); url != "" {
 		return e.NewEndpointFromCloneURL(operation, url)
 	}
 
+	gitDir, err := git.GitDir()
+        if err == nil {
+                url, err := parseFetchHead(strings.Join([]string{gitDir, "FETCH_HEAD"}, "/"));
+                if err == nil {
+                    endpoint := e.NewEndpointFromCloneURL("download", url);
+                    return endpoint
+                } else {
+                    tracerx.Printf("failed parsing FETCH_HEAD: %s", err)
+                }
+        }
+
 	return lfshttp.Endpoint{}
 }
 
+func parseFetchHead(filePath string) (string, error) {
+	file, err := os.Open(filePath)
+	if err != nil {
+		return "", err
+	}
+	defer file.Close()
+
+	scanner := bufio.NewScanner(file)
+	if scanner.Scan() {
+		line := scanner.Text()
+		return ExtractRemoteUrl(line)
+	}
+
+	return "", fmt.Errorf("Failed to read content from %s", filePath)
+}
+
+func ExtractRemoteUrl(line string) (string, error) {
+        re := regexp.MustCompile(`\s(https://|ssh://)?([\/\.\-\:\_a-zA-Z0-9]+)$`)
```

Comment:
How does this work for local fetches with local paths and `file:///` URLs?

Reply:
interesting, I tested this with a local checkout:
```
# make a local repo with an lfs file
mkdir lfs-local-src && cd lfs-local-src
git init .
fallocate -l 10000000 big.chung
git lfs track --filename big.chung
git add .
git commit -m "add big file"

# make another local repo, fetch from the other repo via `file://`
mkdir ~/lfs-local && cd ~/lfs-local
git init .
git fetch file:///home/bmc/sources/lfs-local-src refs/heads/master
GIT_TRACE=1 GIT_TRANSFER_TRACE=1 GIT_CURL_VERBOSE=1 git archive 76e14b4d94fed6af4cb1243adf6f3fdcf5e73077
# git archive output:
# pax_global_header00006660000000000000000000000064144602460010014507gustar00rootroot0000000000000052 comment=7d4efb35e1db865d69edf7aa254ab8c6e93fedac
# big.chung000066400000000000000000000002051446024600100126370ustar00rootroot00000000000000version https://git-lfs.github.com/spec/v1
# oid sha256:f5e02aa71e67f41d79023a128ca35bad86cf7b6656967bfe0884b3a3c4325eaf
# size 10000000
```
and it worked fine! but I agree with your implication that `ExtractRemoteUrl` should not work here, trying to figure out what other codepath must have been triggered